segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1435 @@
+     * @param data    An InputStream representing the data to write.
+     * @param length  The length of the InputStream.
+     * @param timeout Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will indicate the operation succeeded. If the operation failed,
+     * it will contain the cause of the failure. Notable exceptions:
+     * <ul>
+     * <li> BadOffsetException: When the given offset does not match the actual length of the segment in storage.
+     * <li> StreamSegmentNotExistsException: When the given Segment does not exist in Storage.
+     * <li> StorageNotPrimaryException: When this Storage instance is no longer primary for this Segment (it was fenced out).
+     * </ul>
+     * @throws IllegalArgumentException If handle is read-only.
+     */
+    @Override
+    public CompletableFuture<Void> write(SegmentHandle handle, long offset, InputStream data, int length, Duration timeout) {
+        checkInitialized();
+        return execute(() -> {
+            long traceId = LoggerHelpers.traceEnter(log, "write", handle, offset, length);
+            Timer timer = new Timer();
+
+            // Validate preconditions.
+            Preconditions.checkArgument(null != handle, "handle");
+            Preconditions.checkArgument(null != data, "data");
+            String streamSegmentName = handle.getSegmentName();
+            Preconditions.checkArgument(null != streamSegmentName, "streamSegmentName");
+            Preconditions.checkArgument(!handle.isReadOnly(), "handle");
+            Preconditions.checkArgument(offset >= 0, "offset");
+            Preconditions.checkArgument(length >= 0, "length");
+
+            ArrayList<SystemJournal.SystemJournalRecord> systemLogRecords = new ArrayList<>();
+            List<ChunkNameOffsetPair> newReadIndexEntries = new ArrayList<ChunkNameOffsetPair>();
+            int chunksAddedCount = 0;
+            boolean isCommited = false;
+
+            try (MetadataTransaction txn = metadataStore.beginTransaction()) {
+                boolean didSegmentLayoutChange = false;
+
+                // Retrieve metadata.
+                SegmentMetadata segmentMetadata = (SegmentMetadata) txn.get(streamSegmentName);
+
+                // Validate preconditions.
+                checkSegmentExists(streamSegmentName, segmentMetadata);
+
+                segmentMetadata.checkInvariants();
+
+                if (segmentMetadata.isSealed()) {
+                    throw new StreamSegmentSealedException(streamSegmentName);
+                }
+
+                if (segmentMetadata.getOwnerEpoch() > this.epoch) {
+                    throw new StorageNotPrimaryException(streamSegmentName);
+                }
+
+                boolean isSystemSegment = isStorageSystemSegment(segmentMetadata);
+
+                // Check if this is a first write after ownership changed.
+                boolean isFirstWriteAfterFailover = segmentMetadata.isOwnershipChanged();
```

Comment:
All changes to metadata are done in transaction. Transaction will fail to commit if new epoch has claimed ownership and fenced this instance. (Transaction commit writes to the table segment which writes to tier-1 and it will fail because of fencing.)

1. In case of transaction failure, the newly added chunks are deleted. The metadata in table is not updated as transaction failed.
2. On fencing, the new instance will start a new chunk. And update length of old segment by reading length from LTS. So any data written by this instance to existing chunks is simply ignored.